### PR TITLE
Work In Progress: Fix #2

### LINKED ISSUE #2: Bad location data:
UPDATE: See #3 for the fix.

``` js
var acorn = require('acorn');
require('acorn-es7-plugin')(acorn);

var tokens = [];

var ast = acorn.parse(
  [
    'async function myFunc() {',
    '  assert((await somePromise(3)) === "foo");',
    '}'
  ].join('\n'),

  {
    sourceType: 'module',
    ecmaVersion: 7,
    locations: true,
    ranges: true,
    onToken: tokens,
    plugins: {asyncawait: true}
  }
);

console.log(JSON.stringify(ast, null, 4));
```

relevant portion of the AST:

``` json
{
    "type": "AwaitExpression",
    "start": 42,
    "end": 56,
    "loc": {
        "start": {
            "line": 2,
            "column": 16
        },
        "end": {
            "start": {
                "line": 2,
                "column": 16
            },
            "end": {
                "line": 2,
                "column": 30
            }
        }
    },
    "range": [
        42,
        56
    ],
    "operator": "await",
    "argument": {
        "type": "CallExpression",
        "start": 42,
        "end": 56,
        "loc": {
            "start": {
                "line": 2,
                "column": 16
            },
            "end": {
                "line": 2,
                "column": 30
            }
        },
        ...
    }
}
```

### PR DESCRIPTION
**Pull request by jamestalmage**
@MatAtBread 
This is a work in progress fix for #2. Not ready for a merge (see TODO comments).

Any early feedback is appreciated.

### CODE REVIEW

**Comment by jamestalmage** [comment]
@MatAtBread 
Should be ready for your review.

See remaining TODO comments.

I will squish when you are ready to merge.

File: acorn-es7-plugin.js [hunk 1]
```diff
@@ -72,6 +76,9 @@ function asyncAwaitPlugin (parser,options){
 					this.next() ;
 					var r = this.parseStatement(declaration, topLevel) ;
 					r.async = true ;
+					// TODO: This is probably required, but I don't understand nodent well enough to write a test case
+					// r.start = start;
+					// r.loc && (r.loc.start = startLoc);
```
Comment (jamestalmage) [inline acorn-es7-plugin.js:81]: I am guessing these lines should be uncommented. It is for your `NON-STANDARD EXTENSION` for `options.asyncExits` for enodent.

Tell me how you want me to handle that.

File: test/test.js [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    beforeEach(() => {
+      code = [
+        'var x = {',
+        '  async foo() {}',
+        '};'
+      ];
+      node = find(
+        // TODO: Is it really supposed to mark the Property async? Why not the FunctionExpression?
+        'Property',
+        parse(code)
```
Comment (jamestalmage) [inline test/test.js:151]: I find it surprising the `async` modifier gets placed on the `Property` and not the `FunctionExpression`. Are you sure that is right? (I'm guessing you know what you are doing :smile:).

**Comment by MatAtBread** [comment]
At first glance, it looks good. I'll merge and test all my stuff, and commit/release as soon as I get a chance

File: test/test.js (hunk 2 above)
Comment (MatAtBread) [inline test/test.js:151]: Yeah, that's almost certainly a bug. I'll take a look. Excellent catch.

**Comment by MatAtBread** [comment]
I've re-commited 1.0.9 with your fixes in (thanks!). I've also moved all the test installation into it's own package.json - this is a personal preference as I don't like to install babel/mocha/etc on every machine I'm running (I do analytics for a major publisher, and it's a lot!), `npm test` installs it's own devDependencies at test time, not installation time, so you shouldn't see any difference, and the test still passes.

FYI, the `prop.async = true` isn't a bug, the prop (in this case) is of the form `async fn() { ... }` within a class or object definition, and actually represents the method (which is async).

**Comment by jamestalmage** [comment]
> the prop (in this case) is of the form `async fn() { ... }` within a class or object definition, and actually represents the method (which is async)

Yes, I get that. My question is why isn't it `prop.value.async = true`

When we parse the following:

``` js
var foo = {
  async bar() {...}
}
```

Your plugin produces the following AST:

``` js
{
  type: 'ObjectExpression'
  properties: [
    {
      type: 'Property',
      async: true,          // you put this here
      value: {
        type: 'FunctionExpression',
        async: undefined     // I am asking why not put it here?
      }
    }
  ]
}
```

**Comment by MatAtBread** [comment]
Ah. That doesn't look right. I'll look into it.

**Comment by jamestalmage** [comment]
Is there an actual spec for this? I can't find one. The [Mozilla Parser API docs](https://developer.mozilla.org/en-US/docs/Mozilla/Projects/SpiderMonkey/Parser_API) make no mention of `async`.

**Comment by MatAtBread** [comment]
All Functions can be async (https://github.com/estree/estree/blob/master/experimental/async-functions.md), and MemberDefinitions have a value defined as a FuinctionExpression <: Function. So it's definitely (as we thought) a bug

**Comment by MatAtBread** [comment]
I've patched this and will retest tomorrow. I've not yet published, so I'll stick at 1.0.9, but it was definitely wrong, so I'll hopefully get it all working together and so you should be good to go.

**Pull request merged**